src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -84,14 +84,132 @@ private static IEnumerable<TextEdit> FormatRazor(FormattingContext context, Razo
             {
                 // Disclaimer: CSharpCodeBlockSyntax is used a _lot_ in razor so these methods are probably
                 // being overly careful to only try to format syntax forms they care about.
-                TryFormatCSharpCodeBlock(context, edits, source, node);
+                TryFormatCSharpBlockStructure(context, edits, source, node);
                 TryFormatSingleLineDirective(context, edits, source, node);
+                TryFormatBlocks(context, edits, source, node);
             }
 
             return edits;
         }
 
-        private static void TryFormatCSharpCodeBlock(FormattingContext context, List<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        private static bool TryFormatBlocks(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            return TryFormatFunctionsBlock(context, edits, source, node) ||
+                TryFormatCSharpExplicitTransition(context, edits, source, node) ||
+                TryFormatHtmlInCSharp(context, edits, source, node) ||
+                TryFormatComplexCSharpBlock(context, edits, source, node);
+        }
+
+        private static bool TryFormatFunctionsBlock(FormattingContext context, IList<TextEdit> edits, RazorSourceDocument source, SyntaxNode node)
+        {
+            // @functions
+            // {
+            // }
+            if (node is CSharpCodeBlockSyntax directiveCode &&
+                directiveCode.Children.Count == 1 && directiveCode.Children.First() is RazorDirectiveSyntax directive &&
+                directive.Body is RazorDirectiveBodySyntax directiveBody &&
+                directiveBody.Keyword.GetContent().Equals("functions"))
+            {
+                var cSharpCode = directiveBody.CSharpCode;
+                var openBrace = cSharpCode.Children.FirstOrDefault(c => c.Kind == SyntaxKind.RazorMetaCode && c.GetContent().Equals("{", StringComparison.Ordinal));
+                var closeBrace = cSharpCode.Children.LastOrDefault(c => c.Kind == SyntaxKind.RazorMetaCode && c.GetContent().Equals("}", StringComparison.Ordinal));
```

Comment:
There is a `TryGetOpenBrace` local function in this file, perhaps it can be made into a full method and used here? I don't know why exactly but the call to `GetContent()` seems scary to me from a perf point of view. Though I'm sure all of formatting has terrible perf anyway

Reply:
+1 on GetContent seeming super scary. I would hope we could look at token types to find a open/close brace